File: app-tracker-automation/components/base_component.py
"""
Base Component class for reusable UI components.
Provides common functionality for all component objects.
"""

from playwright.sync_api import Page, expect
from utils.logger import Logger
from utils.waits import Waits


class BaseComponent:
    """Base class for all UI components"""
    
    def __init__(self, page: Page, locator: str):
        self.page = page
        self.locator = locator
        self.logger = Logger()
        self.waits = Waits(page)
        self.element = page.locator(locator)
    
    def is_visible(self) -> bool:
        """Check if component is visible"""
        return self.element.is_visible()
    
    def is_enabled(self) -> bool:
        """Check if component is enabled"""
        return self.element.is_enabled()
    
    def wait_for_visible(self, timeout: int = 30000):
        """Wait for component to be visible"""
        self.waits.wait_for_element_visible(self.locator, timeout)
    
    def wait_for_enabled(self, timeout: int = 30000):
        """Wait for component to be enabled"""
        self.waits.wait_for_element_enabled(self.locator, timeout)
    
    def click(self):
        """Click on component"""
        self.element.click()
    
    def hover(self):
        """Hover over component"""
        self.element.hover()
    
    def get_text(self) -> str:
        """Get text content of component"""
        return self.element.text_content()
    
    def get_attribute(self, attribute: str) -> str:
        """Get attribute value of component"""
        return self.element.get_attribute(attribute)
    
    def scroll_into_view(self):
        """Scroll component into view"""
        self.element.scroll_into_view_if_needed()

File: ts-framework/components/base_component.ts
import { Page, Locator } from '@playwright/test';

/**
 * Base Component class for reusable UI components.
 * Provides common functionality for all component objects.
 */
export class BaseComponent {
    protected page: Page;
    protected locator: string;
    protected element: Locator;

    constructor(page: Page, locator: string) {
        this.page = page;
        this.locator = locator;
        this.element = page.locator(locator);
    }

    /**
     * Check if component is visible
     */
    async isVisible(): Promise<boolean> {
        return await this.element.isVisible();
    }

    /**
     * Check if component is enabled
     */
    async isEnabled(): Promise<boolean> {
        return await this.element.isEnabled();
    }

    /**
     * Wait for component to be visible
     */
    async waitForVisible(timeout: number = 30000): Promise<void> {
        await this.element.waitFor({ state: 'visible', timeout });
    }

    /**
     * Wait for component to be enabled
     */
    async waitForEnabled(timeout: number = 30000): Promise<void> {
        await this.element.waitFor({ state: 'visible', timeout });
        // Playwright doesn't have a direct 'enabled' state in waitFor, 
        // but we can use expect or a custom check if needed.
    }

    /**
     * Click on component
     */
    async click(): Promise<void> {
        await this.element.click();
    }

    /**
     * Hover over component
     */
    async hover(): Promise<void> {
        await this.element.hover();
    }

    /**
     * Get text content of component
     */
    async getText(): Promise<string | null> {
        return await this.element.textContent();
    }

    /**
     * Get attribute value of component
     */
    async getAttribute(attribute: string): Promise<string | null> {
        return await this.element.getAttribute(attribute);
    }

    /**
     * Scroll component into view
     */
    async scrollIntoView(): Promise<void> {
        await this.element.scrollIntoViewIfNeeded();
    }
}
